File: prob2/validate_pass.c
#include <stdio.h>
#include <stdint.h>
#include <assert.h>

// Password requirements
#define REQUIRED_UPPER_CASE (1)
#define REQUIRED_NUMBERS    (1)
#define REQUIRED_SPECIAL    (1)

// Number of supported special chars
#define NUM_SPECIAL         (6)       

// Global array of supported special chars
char special[NUM_SPECIAL] = {'!', '@', '#', '$', '%', '&'};

// returns true if c is uppercase
uint8_t is_upper(char c) {
    return ('A' <= c) && (c <= 'Z');
}

// returns true if c is lowercase
uint8_t is_lower(char c) {
    // TODO: check for lowercase character
    return 0;
}

/* returns true if c is in the list of 
 *   special characters */
uint8_t is_special(char c) {
    // TODO: iterate over the special array
    if (c == special[0]) {
        return 1;
    } else {
        return 0;
    }
}

// returns true if c is a number
uint8_t is_number(char c) {
    // TODO: check if c is a number character
    return 0;
}


#define NUM_TESTS (2)

// custom password test type
typedef struct password_test_t {
    const char* password;
    uint8_t length;
    uint8_t result;
} password_test_t;

// array of password tests
static password_test_t password_tests[NUM_TESTS] = {{"IMeet3Requirements!", 21, 1},
                                                        // only numbers
                                                        {"12345678", 9, 0}};


// validates the password argument against password requirement macros
uint8_t validate_password(const char* password, uint8_t length) {

    uint8_t upper_count = 0;
    uint8_t special_count = 0;
    uint8_t number_count = 0;

    // Iterate over each char in the password
    for (int i = 0; i < length; i++) {
        // The \0 character is special and marks the end of a string
        if (password[i] == '\0') {
            break;
        }
        if (is_upper(password[i])) {
            // ++ is a special operator that is equivalent to upper_count = upper_count + 1;
            upper_count++;
        } else if (is_lower(password[i])) {
            // no-op
        } else if (is_special(password[i])) {
            special_count++;
        } else if (is_number(password[i])) {
            number_count++;
        }
        else {
            printf("Unsupported character!\n");
            return 0;
        }
    }

    // Check that our counts meet the requirements
    // TODO: check other requirements
    return special_count >= REQUIRED_SPECIAL;
}

int main(int argc, char* argv[]) {
    for (int i = 0; i < NUM_TESTS; i++) {
        assert(validate_password(password_tests[i].password, password_tests[i].length) == password_tests[i].result);
    }
    printf("All tests passed!\n");
    return 0;
}

File: prob2/validate_pass.ml
(* returns true iff c is uppercase *)
  let is_upper (c : char) : bool =
    c >= 'A' && c <= 'Z'
  
  (* returns true iff c is lowercase *)
  let is_lower (c : char) : bool =
    false
  
  (* returns true iff c is in s *)
  let rec is_special_generic (s : char list) (c : char) : bool =
    match s with
    | [] -> false
    | hd :: tl -> 
      if hd = c then true 
      else (* TODO: recursively call is_special_generic on the rest of the list *) false
  
  (* list of special characters *)
  let specials = ['!'; '@'; '#'; '$'; '%'; '&']
  (* function that takes a char *)
  let is_special = is_special_generic specials
  
  (* returns true iff c is a number *)
  let is_number (c : char) : bool =
    false
  
  (* custom type for password tests *)
  type password_test = PasswordTest of string * bool
  
  (* given the password requirements returns true iff s meets the requirements *)
  let validate_pass_generic (numbers : int) (upper : int) (special : int) (s : string) : bool =
    let upper_count = String.fold_left (fun a c -> if is_upper c then a + 1 else a) 0 s in
    let special_count = String.fold_left (fun a c -> if is_special c then a + 1 else a) 0 s in
    let number_count = String.fold_left (fun a c -> if is_number c then a + 1 else a) 0 s in
    let lower_count  = String.fold_left (fun a c -> if is_lower c then a + 1 else a) 0 s in
    let unsupported = (String.length s - (upper_count + special_count + lower_count + number_count)) <> 0 in
    (* TODO: check uppercase and special *)
    number_count >= numbers && not unsupported
  
  (* our requirements for a password *)
  let required_numbers   = 1
  let required_special   = 1
  let required_uppercase = 1
  (* function that takes a string to validate *)
  let validate_pass = validate_pass_generic required_numbers required_uppercase required_special
  
  (* given a list of tests, ensures they all pass *)
  let rec test_passwords (tests : password_test list) : unit =
    match tests with
    | [] -> ()
    | PasswordTest (s, res) :: tl ->
      assert (validate_pass(s) = res);
      test_passwords tl
  
  let run_tests () : unit =
    let test1 : password_test = PasswordTest ("IMeet3Requirements!", true) in
    let test2 : password_test = PasswordTest ("12345678", false) in
    (* TODO: add additional test cases *)
    let password_tests = [test1; test2] in
    test_passwords password_tests
  
  
  let _ = run_tests () ;;